pg8000: Support range and multirange types

tlocke opened the pull request:
### Description

From version 1.29.7, pg8000 supports `range` and `multirange` types. This change to the pg8000 dialect enables these range types in SQLAlchemy.

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [x ] A short code fix
	- This fixes https://github.com/sqlalchemy/sqlalchemy/issues/9965
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

tlocke committed:
pg8000: Support range and multirange types

Fixes: #9965

tlocke commented:
One thing I found is that as it stands there are some tests that are skipped, but would actually pass. It's the tests in https://github.com/sqlalchemy/sqlalchemy/blob/main/test/dialect/postgresql/test_types.py that use `_RangeTypeRoundTrip` and that are marked `@testing.requires.any_psycopg_compatibility`. The problem is that lifting this requirement then enables other tests that pg8000 fails on. So any help on the right thing to do here is greatly appreciated!

tlocke commented:
The other thing where I'd be grateful for some advice is on the dbapi module that the dialect uses. So at the moment pg8000 has three APIs:

- **pg8000.dbapi** - This strictly follows the DBAPI-2 standard
- **pg8000.native** - This breaks with the DBAPI standard and is designed to be more Pythonic.
- **pg8000** - This is the legacy API, which SQLAlchemy currently uses, but which I'm trying to deprecate because it's neither fish nor fowl, ie. it doesn't strictly follow DBAPI-2, and also doesn't have the freedom to be something different.

This PR moves the API from ``pg8000`` to the ``pg8000.dbapi`` API. In the patch I have the line:

`        pg8000_Range = dialect.dbapi.dbapi.Range`

but I don't understand why it didn't work to just do:

`        pg8000_Range = dialect.dbapi.Range`

CaselIT commented:
Thanks,

I'll take a better look later, just a few initial comments

> testing.requires.any_psycopg_compatibility

Feel free to add another requirement there. Like pg_ranges or similar. Should go in the test requirements only since it's not used in the suite 

> So at the moment pg8000 has three APIs:

Does this need a version check? Meaning is that supported also from older pg8000 versions?

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision f2b65942553a3752fc16278cb7f336fdeac56241 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Failed to create a gerrit review, git squash against branch 'main' failed

zzzeek commented:
can you rebase this against the main branch?  then we can push into gerrit, thanks

tlocke commented:
> can you rebase this against the main branch? then we can push into gerrit, thanks

I've rebased, so it should be okay for gerrit now.

> testing.requires.any_psycopg_compatibility

I've now changed these to:

`@testing.requires.psycopg_or_pg8000_compatibility`

so that problem should be solved now.

> > So at the moment pg8000 has three APIs:
> 
> Does this need a version check? Meaning is that supported also from older pg8000 versions?

The creation of the three separate APIs happened at version 1.17.0 which was release on 2021-01-30. Not sure if this warrants a version check, but I can put one in if needed?

zzzeek commented:
> > Does this need a version check? Meaning is that supported also from older pg8000 versions?
> 
> The creation of the three separate APIs happened at version 1.17.0 which was release on 2021-01-30. Not sure if this warrants a version check, but I can put one in if needed?

we usually only need version checks for behaviors that are likely to take place implicitly when using the dialect.  if this is only the "range" feature that right now just fails in any case, the version check is typically not warranted.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 11076884c6b37766cabe98621383935fa2dc6c86 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
New Gerrit review created for change 11076884c6b37766cabe98621383935fa2dc6c86: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4730

zzzeek commented on lib/sqlalchemy/dialects/postgresql/pg8000.py:
```diff
@@ -409,7 +486,7 @@ def _dbapi_version(self):
     @classmethod
     def import_dbapi(cls):
-        return __import__("pg8000")
+        return __import__("pg8000.dbapi")
```
this should be like this:

     return __import__("pg8000.dbapi").dbapi

python thing

zzzeek commented:
> This PR moves the API from `pg8000` to the `pg8000.dbapi` API. In the patch I have the line:
> 
> ` pg8000_Range = dialect.dbapi.dbapi.Range`
> 
> but I don't understand why it didn't work to just do:
> 
> ` pg8000_Range = dialect.dbapi.Range`

``__import__('a.b.c.')`` returns "a", so you have to drill back down to `.b.c` to get the endpoint

zzzeek commented:
> The creation of the three separate APIs happened at version 1.17.0 which was release on 2021-01-30. Not sure if this warrants a version check, but I can put one in if needed?

our bottom version for pg8000 is set to 1.29.1 in setup.cfg so we should be fine

tlocke commented:
Actually I've found some problems with switching away from the pg8000 legacy API, so I think it's best to avoid making that change in this pull request. I've also noticed a bug in pg8000 1.29.7, so that's fixed and we're on 1.29.8 now.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision f69dbff668682f0cdbfa8bb72672707bd40e1bd6 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Failed to create a gerrit review, git squash against branch 'main' failed

CaselIT commented:
@zzzeek is seems that the bot has now more difficulty on updating the gerrit changes. IIRC it never happened since a few weeks ago. Is something changed?

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision 7b7cdb2c107c95e76fa3a9c50942a397e76f8c48 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset 7b7cdb2c107c95e76fa3a9c50942a397e76f8c48 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4730

CaselIT commented:
> Actually I've found some problems with switching away from the pg8000 legacy API, so I think it's best to avoid making that change in this pull request. I've also noticed a bug in pg8000 1.29.7, so that's fixed and we're on 1.29.8 now.

just to confirm, for now the driver in sqlalchemy will continue to use the toplevel module, correct?

tlocke commented:
Yes that's right.

sqla-tester commented:
**mike bayer** (zzzeek) wrote:

looks good, if we can add a changelog file we'll be set

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4730

tlocke commented:
> looks good, if we can add a changelog file we'll be set

I've added in a brief changelog file, let me know if I should add anything else to it.

sqla-tester reviewed (commented):
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision b48dc3ef9cbb267fe945d08ca38240c68b716f79 of this pull request into gerrit so we can run tests and reviews and stuff

sqla-tester commented:
Patchset b48dc3ef9cbb267fe945d08ca38240c68b716f79 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4730

sqla-tester commented:
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4730 has been **merged**. Congratulations! :)
